docs/understanding-airbyte/airbyte-protocol.md

```diff
@@ -0,0 +1,735 @@
+        "properties": {
+          "name": {
+            "type": "string"
+          },
+          "age": {
+            "type": "number"
+          }
+        }
+      }
+    }
+  ]
+}
+```
+If we are using a data source that wraps an API with multiple different resources (e.g. `api/customers` and `api/products`) each route would correspond to a stream. The JSON object returned by each route would be described in the `json_schema` field.
+
+e.g. In the case where the API has two endpoints `api/customers` and `api/products` and each returns a list of JSON objects, the `AirbyteCatalog` might look like this. (Note: using the JSON schema standard for defining a stream allows us to describe nested objects. We are not constrained to a classic "table/columns" structure)
+
+```json
+{
+  "streams": [
+    {
+      "name": "customers",
+      "json_schema": {
+        "type": "object",
+        "required": ["name"],
+        "properties": {
+          "name": {
+            "type": "string"
+          }
+        }
+      }
+    },
+    {
+      "name": "products",
+      "json_schema": {
+        "type": "object",
+        "required": ["name", "features"],
+        "properties": {
+          "name": {
+            "type": "string"
+          },
+          "features": {
+            "type": "array",
+            "items": {
+              "type": "object",
+              "required": ["name", "productId"],
+              "properties": {
+                "name": { "type": "string" },
+                "productId": { "type": "number" }
+              }
+            }
+          }
+        }
+      }
+    }
+  ]
+}
+```
+
+**Note:** Stream and field names can be any UTF8 string. Destinations are responsible for cleaning these names to make them valid table and column names in their respective data stores.
+
+## Namespace
+Technical systems often group their underlying data into namespaces with each namespace's data isolated from another namespace. This isolation allows for better organisation and flexibility, leading to better usability.
+
+An example of a namespace is the RDMS's `schema` concept. Some common use cases for schemas are enforcing permissions, segregating test and production data and general data organisation.
+
+The `AirbyteStream` represents this concept through an optional field called `namespace`. Additional documentation on Namespaces can be found [here](#namespace).
+
+
+## Cursor
+
+* The cursor is how sources track which records are new or updated since the last sync.
+* A "cursor field" is the field that is used as a comparable for making this determination.
+    * If a configuration requires a cursor field, it requires an array of strings that serves as a path to the desired field. e.g. if the structure of a stream is `{ value: 2, metadata: { updated_at: 2020-11-01 } }` the `default_cursor_field` might be `["metadata", "updated_at"]`.
+
+## AirbyteStream
+
+This section will document the meaning of each field in an `AirbyteStream`
```

Comment:
It feels like these should be comments directly on the definition of `AirbyteStream` embedded here. It seems like an antipattern to keep the explanation of fields separate from its actual YAML code defiition

Reply:
agreed. i think that'll be the next iteration.